resources/test/tests/unit/promise_setup.html

```diff
@@ -0,0 +1,333 @@
+          });
+        }, 'after');
+      }
+    ).then(({harness, tests}) => {
+      assert_equals(harness, 'OK');
+      assert_equals(tests.after, 'PASS');
+    });
+}, 'Honors configuration properties');
+
+promise_test(() => {
+  return makeTest(
+      () => {
+        // Ensure that the harness error is the result of explicit error
+        // handling
+        setup({ allow_uncaught_exception: true });
+
+        test(() => {}, 'before');
+        promise_setup(() => { throw new Error('this error is expected'); });
+        promise_test(() => Promise.resolve(), 'after');
+      }
+    ).then(({harness, tests}) => {
+      assert_equals(harness, 'ERROR');
+      assert_equals(tests.before, 'PASS');
+      assert_equals(tests.after, 'NOTRUN');
+    });
+}, 'Error for synchronous exceptions');
+
+promise_test(() => {
+  return makeTest(
+      () => {
+        // Ensure that the harness error is the result of explicit error
+        // handling
+        setup({ allow_uncaught_exception: true });
+
+        test(() => {}, 'before');
+        promise_setup(() => undefined);
+        promise_test(() => Promise.resolve(), 'after');
+      }
+    ).then(({harness, tests}) => {
+      assert_equals(harness, 'ERROR');
+      assert_equals(tests.before, 'PASS');
+      assert_equals(tests.after, 'NOTRUN');
+    });
+}, 'Error for missing return value');
+
+promise_test(() => {
+  return makeTest(
+      () => {
+        // Ensure that the harness error is the result of explicit error
+        // handling
+        setup({ allow_uncaught_exception: true });
+
+        test(() => {}, 'before');
+        var noThen = Promise.resolve();
+        noThen.then = undefined;
+        promise_setup(() => noThen);
+        promise_test(() => Promise.resolve(), 'after');
+      }
+    ).then(({harness, tests}) => {
+      assert_equals(harness, 'ERROR');
+      assert_equals(tests.before, 'PASS');
+      assert_equals(tests.after, 'NOTRUN');
+    });
+}, 'Error for non-thenable return value');
+
+promise_test(() => {
+  return makeTest(
+      () => {
+        // Ensure that the harness error is the result of explicit error
+        // handling
+        setup({ allow_uncaught_exception: true });
+
+        test(() => {}, 'before');
+        var poisonedThen = {
+          get then() {
+            throw new Error('this error is expected');
+          }
+        };
+        promise_setup(() => poisonedThen);
+        promise_test(() => Promise.resolve(), 'after');
+      }
+    ).then(({harness, tests}) => {
+      assert_equals(harness, 'ERROR');
+      assert_equals(tests.before, 'PASS');
+      assert_equals(tests.after, 'NOTRUN');
+    });
+}, 'Error for "poisoned" `then` property');
+
+promise_test(() => {
+  return makeTest(
+      () => {
+        // Ensure that the harness error is the result of explicit error
+        // handling
+        setup({ allow_uncaught_exception: true });
+
+        test(() => {}, 'before');
+        var badThen = {
```

Comment:
Dumb question: how does this differ semantically from the poisonedThen case?

Reply:
The implementation could get the property value without triggering the accessor
by using API's like `getOwnPropertyDescriptor`. An edge case, to be sure, but I
couldn't help myself.